lib/addressable/idna/native.rb

```diff
@@ -29,10 +29,6 @@ def self.punycode_decode(value)
        IDN::Punycode.decode(value.to_s)
      end
 
-    def self.unicode_normalize_kc(value)
-      IDN::Stringprep.nfkc_normalize(value.to_s)
-    end
```

Comment:
@jarthod Maybe we were a bit aggressive removing the method `Addressable::IDNA.unicode_normalize_kc` in a patch version. There might be code in the wild using these methods. Just happened to see [some code using `Addressable::IDNA.to_ascii`](https://github.com/discourse/discourse/pull/17840/files#diff-f98dab56f29862edb3b9003dde7224a286c23262bef7328698c4864c4ade1c6fR157).

Do you think it is possible to add it back? (In both native and pure versions, we don't know what people are using)

Reply:
Would be cool to mark it as deprecated (though I don't think this project as an established way of doing that, some new code would be needed)

Reply:
Why not, as you prefer. We can add it back easily, yes, with some caveats:
1. For `Pure` I don't think we should restore the old ruby implementation but use the new one (ruby's implementation) so it won't be 100% compatible with before but at least the method will exist.
2. Same thing for `Native`, if we restore the `IDN::Stringprep` one (meaning 100% identical to before the removal) the result will now be different than what addressable uses (ruby normalize). Also there's [no equivalent in `libidn2` AFAICS](https://libidn.gitlab.io/libidn2/reference/api-index-full.html). So we should also forward to ruby normalize I suppose.

---

For the deprecation, I didn't see any existing way either in the project. From a quick look I found this std ruby module: https://ruby-doc.org/stdlib-3.1.0/libdoc/rubygems/rdoc/Gem/Deprecate.html
Which seems to be [recommended](https://stackoverflow.com/a/23554720/304434) to do that (never used it). It prints this standard message we saw many times, would look like this (manually replacing the interpolations):
```
NOTE: Addressable::IDNA.unicode_normalize_kc is deprecated; use String#unicode_normalize(:nfkc) instead. It will be removed on or after 2023-04.
Addressable::IDNA.unicode_normalize_kc called from xxxxxxxx:42
```
The date is not optional unfortunately, but the message says "or after" so it's not a big deal we can leave "2023-04" so people know it's deprecated and will stop working at some point. This message allows to provide the replacement code which is good. Also it support the _standard_ `Gem::Deprecate.skip` option.

Alternatively we can [use `Kernel.warn` directly](https://stackoverflow.com/a/294114/304434), which even has a ["category" option which can take the value `:deprecated`](https://ruby-doc.org/3.2.1/Kernel.html#method-i-warn), though it's not clear what it changes and even Gem::Deprecate doesn't use it.

Reply:
Thanks yet again for an awesome comment @jarthod :) I agree with you that we should not restore the old behaviours, just bring back the methods so code doesn't break. I'm fine with either options for deprecating.

Reply:
→ #504